Sources/SocksCore/TCPSocket.swift

```diff
@@ -76,45 +58,84 @@ public class TCPSocket: InternetSocket {
         }
         return buffer
     }
+}
+
+extension TCPWriteableSocket {
     
     public func send(data: [UInt8]) throws {
         let len = data.count
-        let flags: Int32 = 0 //FIXME: allow setting flags with a Swift enum
+        let flags = Int32(SOCKET_NOSIGNAL) //FIXME: allow setting flags with a Swift enum
         let sentLen = socket_send(self.descriptor, data, len, flags)
-        guard sentLen == len else { throw Error(.SendFailedToSendAllBytes) }
+        guard sentLen == len else { throw Error(.sendFailedToSendAllBytes) }
+    }
+}
+
+public class TCPInternetSocket: InternetSocket, TCPSocket, TCPReadableSocket, TCPWriteableSocket {
+
+    public let descriptor: Descriptor
+    public let config: SocketConfig
+    public let address: ResolvedInternetAddress
+    public var closed: Bool
+
+    public required init(descriptor: Descriptor?, config: SocketConfig, address: ResolvedInternetAddress) throws {
+        if let descriptor = descriptor {
+            self.descriptor = descriptor
+        } else {
+            self.descriptor = try TCPInternetSocket.createNewSocket(config: config)
+        }
+        self.config = config
+        self.address = address
+        self.closed = false
+        
+        self.reuseAddress = true
+    }
+    
+    public convenience init(address: InternetAddress) throws {
+        var conf: SocketConfig = .TCP(addressFamily: address.addressFamily)
+        let resolved = try address.resolve(with: conf)
+        try conf.adjust(for: resolved)
```

Comment:
Very good point, I did something slightly different, but it makes it impossible to forget to adjust it now: https://github.com/czechboy0/Socks/pull/52

Thanks for noticing this